src/Command/PurgeCommand.php

```diff
@@ -0,0 +1,318 @@
+
+class PurgeCommand extends Command
+{
+    /**
+     * @var ContaoFrameworkInterface
+     */
+    private $framework;
+
+    /**
+     * @var Connection
+     */
+    private $db;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var Filesystem
+     */
+    private $fs;
+
+    /**
+     * PurgeCommand constructor.
+     * @param ContaoFrameworkInterface $framework
+     * @param Connection $db
+     * @param LoggerInterface $logger
+     * @param Filesystem|null $fs
+     */
+    public function __construct(
+        ContaoFrameworkInterface $framework,
+        Connection $db,
+        LoggerInterface $logger,
+        Filesystem $fs = null
+    ) {
+        $this->framework = $framework;
+        $this->db = $db;
+        $this->logger = $logger;
+        $this->fs = $fs ? $fs : new Filesystem();
+
+        parent::__construct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('leads:purge')
+            ->setDescription('Purge the leads outsite the configured storage period.');
+    }
+
+    /**
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int|null|void
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->framework->initialize();
+
+        $purged = $this->executeBatchPurge();
+
+        if ($purged) {
+            $logMessage = 'The leads have been purged successfully.';
+        } else {
+            $logMessage = 'No leads to purge.';
+        }
+
+        $logLevel = LogLevel::INFO;
+        $this->logger->log($logLevel, $logMessage, array('contao' => new ContaoContext(__METHOD__, $logLevel)));
+        $output->writeln('<info>'.$logMessage.'</info>');
+    }
+
+    /**
+     * @return bool
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function executeBatchPurge()
+    {
+        $purged = false;
+        $forms = $this->db->fetchAll(
+            "SELECT id, title, leadPeriod, leadPurgeUploads FROM tl_form WHERE leadPeriod != ''"
+        );
+
+        foreach ($forms as $masterForm) {
+
+            $leadPeriodTime = $this->convertTimePeriodToTime($masterForm['leadPeriod']);
+            if (!empty($leads = $this->getAllLeads($masterForm['id'], $leadPeriodTime))) {
+                $deletedUploads = null;
+                if (!empty($leadsData = $this->getAllLeadsData($leads))) {
+                    $this->purgeLeadsData($leadsData, $masterForm);
+
+                    if ($masterForm['leadPurgeUploads']) {
+                        $deletedUploads = $this->purgeUploads($leadsData);
+                    }
+                }
+
+                $this->purgeLeads($leads, $masterForm);
+
+                // Add custom logic
+                if (isset($GLOBALS['TL_HOOKS']['postLeadsPurge']) && is_array($GLOBALS['TL_HOOKS']['postLeadsPurge'])) {
```

Comment:
Again, this has to be an event. Hooks are an old concept and not used anymore for new code.

Reply:
what's best practice to register and use hooks in contao 4.4 because there we cannot use the tagged services 'contao.hook'?

Reply:
You don't need to use hooks. Use events instead. See https://github.com/1up-lab/contao-mailchimp/pull/18 (at least I think that's what @Toflar means ;) ).

Reply:
Ah, but this extension is still compatible with Contao 3...

Reply:
No, the develop branch is contao 4 only.

Reply:
Right, sorry, I haven't noticed.